### PR TITLE
Issue #OS-44: validating presence of signatures

### PR DESCRIPTION
**Pull request by jyotsna-tarento**
This PR includes the following:

1. Interceptor to validate presence of signatures.
2. Middleware to validate presence of signatures.
3. Changes to verify if signature is present for all attributes.
4. As of now validation is handled only for create requests, as update request might require further changes.

### CODE REVIEW

File: java/middleware/registry-middleware/signature-presence-validation/src/main/java/io/opensaber/registry/middleware/impl/SignaturePresenceValidator.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+			if(!rNode.isAnon() && !s.getPredicate().equals(RDF.type)){
+				String predicate = s.getPredicate().toString();
+				if(!signatureAttributes.contains(predicate)){
+					predicateList.add(predicate);
+				}
+			}
+		}
+
+		for(String predicate: predicateList){
+			StmtIterator signatureIter = rdf.listStatements(null, prop, (RDFNode)null);
```
Comment (indrajra) [inline java/middleware/registry-middleware/signature-presence-validation/src/main/java/io/opensaber/registry/middleware/impl/SignaturePresenceValidator.java]: am not a good fan of two iterations. Can you check if we could leverage any other methods in Model to just give us a presence or not? Say, contains() method exists?

File: java/middleware/registry-middleware/signature-presence-validation/src/main/java/io/opensaber/registry/middleware/impl/SignaturePresenceValidator.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+import io.opensaber.registry.middleware.util.RDFUtil;
+import io.opensaber.validators.shex.shaclex.ShaclexValidator;
+import io.opensaber.pojos.ValidationResponse;
+
+public class SignaturePresenceValidator implements Middleware{
+
+	private static final String RDF_DATA_IS_MISSING = "RDF Data is missing!";
+	private static final String RDF_DATA_IS_INVALID = "Data validation failed!";
+	private static final String SCHEMA_IS_NULL = "Schema for validation is missing";
+	private static final String SX_EXPRESSION_IRI = "http://shex.io/ns/shex#expression";
```
[the remark this answers is not in the document]
Comment (jyotsna-tarento) [reply java/middleware/registry-middleware/signature-presence-validation/src/main/java/io/opensaber/registry/middleware/impl/SignaturePresenceValidator.java]: :+1:

File: java/middleware/registry-middleware/signature-presence-validation/src/main/java/io/opensaber/registry/middleware/impl/SignaturePresenceValidator.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+	private static final String SX_EXPRESSIONS_IRI = "http://shex.io/ns/shex#expressions";
+	private static final String SX_PREDICATE_IRI = "http://shex.io/ns/shex#predicate";
+	private static final String FORMAT = "JSON-LD";
```
[the remark this answers is not in the document]
Comment (jyotsna-tarento) [reply java/middleware/registry-middleware/signature-presence-validation/src/main/java/io/opensaber/registry/middleware/impl/SignaturePresenceValidator.java]: :+1:

File: java/middleware/registry-middleware/signature-presence-validation/src/main/java/io/opensaber/registry/middleware/impl/SignaturePresenceValidator.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+		}else if (schemaForCreate == null) {
+			throw new MiddlewareHaltException(SCHEMA_IS_NULL);
+		}else {
+			validateSignature((Model)RDF);
+			return mapData;
+		}
+	}
+
+	public Map<String, Object> next(Map<String, Object> mapData) throws IOException {
+		// TODO Auto-generated method stub
```
[the remark this answers is not in the document]
Comment (jyotsna-tarento) [reply java/middleware/registry-middleware/signature-presence-validation/src/main/java/io/opensaber/registry/middleware/impl/SignaturePresenceValidator.java]: :+1:

File: java/middleware/registry-middleware/signature-presence-validation/src/main/java/io/opensaber/registry/middleware/impl/SignaturePresenceValidator.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+		List<String> attributeList = new ArrayList<String>();
+		Resource shapeResource = ResourceFactory.createResource(shape);
+		RDFNode node = getObjectAfterFilter(shapeResource, SX_EXPRESSION_IRI, signatureConfigModel);
+		RDFNode firstNode = getObjectAfterFilter(node, SX_EXPRESSIONS_IRI, signatureConfigModel);
+		getOtherAttributes(firstNode, signatureConfigModel, attributeList);
+		return attributeList;
+	}
+
+	private void getOtherAttributes(RDFNode firstNode, Model signatureConfigModel, List<String> attributeList){
+		RDFNode secondNode = getObjectAfterFilter(firstNode, RDF.rest.getURI(), signatureConfigModel);
```
[the remark this answers is not in the document]
Comment (jyotsna-tarento) [reply java/middleware/registry-middleware/signature-presence-validation/src/main/java/io/opensaber/registry/middleware/impl/SignaturePresenceValidator.java]: Comments have been added, but it's difficult to name the nodes differently as multiple levels of filters are applied before the attributes are retrieved. getOtherAttributes method has been renamed to addAttributesForShape. It's written recursively as of now. So here the function is giving out attributes for the SignatureShape and not the node. Please review again as few changes have been added and please provide your comments.

File: java/middleware/registry-middleware/signature-presence-validation/src/main/java/io/opensaber/registry/middleware/impl/SignaturePresenceValidator.java [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+	private Schema schemaForCreate;
+	private String registryContext;
+	private String signatureConfigName;
+	private String registrySystemBase;
+	private Model schemaConfig;
+	private List<String> signatureTypes = new ArrayList<String>();
+	
+	public SignaturePresenceValidator(Schema schemaForCreate, String registryContext, String registrySystemBase, String signatureConfigName, Map<String,String> shapeTypeMap, Model schemaConfig) {
+		this.schemaForCreate = schemaForCreate;
```
[the remark this answers is not in the document]
Comment (indrajra) [reply java/middleware/registry-middleware/signature-presence-validation/src/main/java/io/opensaber/registry/middleware/impl/SignaturePresenceValidator.java]: It is realized in the same way for other interceptors, so I get it.

File: java/middleware/registry-middleware/signature-presence-validation/src/main/java/io/opensaber/registry/middleware/impl/SignaturePresenceValidator.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (indrajra) [reply java/middleware/registry-middleware/signature-presence-validation/src/main/java/io/opensaber/registry/middleware/impl/SignaturePresenceValidator.java]: Ok for now - some work could be done to make it better.

File: java/registry-interceptor/src/main/java/io/opensaber/registry/interceptor/SignaturePresenceValidationInterceptor.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+	
+	@Override
+	public void afterCompletion(HttpServletRequest arg0, HttpServletResponse arg1, Object arg2, Exception arg3)
+			throws Exception {
+	}
+
+	@Override
+	public void postHandle(HttpServletRequest arg0, HttpServletResponse arg1, Object arg2, ModelAndView arg3)
+			throws Exception {
+		// TODO Auto-generated method stub
```
[the remark this answers is not in the document]
Comment (indrajra) [reply java/registry-interceptor/src/main/java/io/opensaber/registry/interceptor/SignaturePresenceValidationInterceptor.java]: done.

File: java/registry/src/main/resources/validations_create.shex.sample [hunk 8]
```diff
@@ -41,21 +49,35 @@ teacher:TeacherRoleShape CLOSED {
...
 	a [teacher:NonTeachingAssignmentsForAcademicCalendar] ;
-	teacher:daysOfNonTeachingAssignments @teacher:isADecimal
+	teacher:daysOfNonTeachingAssignments @teacher:isADecimal ;
+	teacher:signatures @teacher:SignatureShape *
+}
+
+teacher:SignatureShape CLOSED {
+	a [sc:RsaSignature2018] ;
+	teacher:signatureFor @teacher:isAnyURI ;
+	teacher:creator @teacher:isACreator ;
```
[the remark this answers is not in the document]
Comment (indrajra) [reply java/registry/src/main/resources/validations_create.shex.sample]: done this myself.

**Pull request merged**